File: incident-service/src/main/java/com/emergency/incident/service/StationService.java
package com.emergency.incident.service;

import com.emergency.incident.dto.CreateStationRequest;
import com.emergency.incident.model.ResponderStation;
import com.emergency.incident.model.StationType;
import com.emergency.incident.repository.ResponderStationRepository;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class StationService {

    private final ResponderStationRepository stationRepository;

    public StationService(ResponderStationRepository stationRepository) {
        this.stationRepository = stationRepository;
    }

    public ResponderStation createStation(CreateStationRequest request) {
        ResponderStation station = ResponderStation.builder()
                .name(request.getName())
                .stationType(request.getStationType())
                .latitude(request.getLatitude())
                .longitude(request.getLongitude())
                .address(request.getAddress())
                .phoneNumber(request.getPhoneNumber())
                .capacity(request.getCapacity())
                .isAvailable(true)
                .currentOccupancy(0)
                .build();

        return stationRepository.save(station);
    }

    public List<ResponderStation> getAllStations() {
        return stationRepository.findAll();
    }

    public List<ResponderStation> getStationsByType(StationType type) {
        return stationRepository.findByStationType(type);
    }

    public ResponderStation getStationById(Long id) {
        return stationRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Station not found with id: " + id));
    }

    public ResponderStation updateStationAvailability(Long id, boolean available) {
        ResponderStation station = getStationById(id);
        station.setIsAvailable(available);
        return stationRepository.save(station);
    }

    public ResponderStation updateStationCapacity(Long id, int capacity, int occupancy) {
        ResponderStation station = getStationById(id);
        station.setCapacity(capacity);
        station.setCurrentOccupancy(occupancy);
        return stationRepository.save(station);
    }
}

File: nodejs-migration/incident-service/src/services/StationService.ts
import { AppDataSource } from '../database/database';
import { ResponderStation, StationType } from '../entities/ResponderStation';
import { CreateStationRequest } from '../dto/incident.dto';

export class StationService {
  private stationRepository = AppDataSource.getRepository(ResponderStation);

  async createStation(request: CreateStationRequest): Promise<ResponderStation> {
    const station = this.stationRepository.create({
      name: request.name,
      stationType: request.stationType,
      latitude: request.latitude,
      longitude: request.longitude,
      address: request.address,
      phoneNumber: request.phoneNumber,
      capacity: request.capacity,
      currentOccupancy: 0,
      isAvailable: true,
    });

    return await this.stationRepository.save(station);
  }

  async getAllStations(): Promise<ResponderStation[]> {
    return await this.stationRepository.find();
  }

  async getStationsByType(type: StationType): Promise<ResponderStation[]> {
    return await this.stationRepository.find({
      where: { stationType: type },
    });
  }

  async getStationById(id: number): Promise<ResponderStation> {
    const station = await this.stationRepository.findOne({
      where: { id },
    });

    if (!station) {
      throw new Error(`Station not found with id: ${id}`);
    }

    return station;
  }

  async updateStationAvailability(
    id: number,
    available: boolean
  ): Promise<ResponderStation> {
    const station = await this.getStationById(id);
    station.isAvailable = available;
    return await this.stationRepository.save(station);
  }

  async updateStationCapacity(
    id: number,
    capacity: number,
    occupancy: number
  ): Promise<ResponderStation> {
    const station = await this.getStationById(id);
    station.capacity = capacity;
    station.currentOccupancy = occupancy;
    return await this.stationRepository.save(station);
  }
}
